### PR TITLE
Remove inventory reservations based on orders

### LINKED ISSUE #2311: Improve reservations clean up cron job
### Description
Existing implementation of reservations clean up mechanism relies on a cron job, which is looking for reservations with sum qty of 0 and removes them. In certain cases this may be undesired, or, if more business events (other than order-related) are tracked in the system.

### Desired state
Reservations cleanup is improved to accommodate additional logic:
* Clean-up cron job is configurable, it may be disabled completely with configuration
* One-time reservations clean up is possible with a dedicated CLI command
* Clean-up logic is updated, to work only with order-related reservations, if other types are implemented by SI they would also be responsible for clearing those.
* Criteria for clean-up is enforced. Before, it was checking only qty and sku, after the change it should also check order_id.

### PR DESCRIPTION
**Pull request by Jeroenll**
Created this PR after my conversation with @ishakhsuvarov on slack.

### Description (*)
Our client relies heavily on the rest API to create/update products, fetch order and create shipments. We recently had some issues where they were calling the create shipment API for the same order multiple times. This resulted in multiple shipments being created (this a whole different problem, let't not discuss that now :slightly_smiling_face: ).
So let's say this happenend for order 1, the source reservations will be:
```
+----------+----------+-------------------------------------------------------------------------+
| sku      | quantity | metadata                                                                |
+----------+----------+-------------------------------------------------------------------------+
| MT014958 | -10      | {"event_type":"order_placed","object_type":"order","object_id":"1"}     |
| MT014958 | 10       | {"event_type":"shipment_created","object_type":"order","object_id":"1"} |
| MT014958 | 10       | {"event_type":"shipment_created","object_type":"order","object_id":"1"} |
+----------+----------+-------------------------------------------------------------------------+
```
When a new order is being placed (order 2) another reservation is placed:
```
+----------+----------+-------------------------------------------------------------------------+
| sku      | quantity | metadata                                                                |
+----------+----------+-------------------------------------------------------------------------+
| MT014958 | -10      | {"event_type":"order_placed","object_type":"order","object_id":"1"}     |
| MT014958 | 10       | {"event_type":"shipment_created","object_type":"order","object_id":"1"} |
| MT014958 | 10       | {"event_type":"shipment_created","object_type":"order","object_id":"1"} |
| MT014958 | -10      | {"event_type":"shipment_created","object_type":"order","object_id":"2"} |
+----------+----------+-------------------------------------------------------------------------+
```
Order 2 is placed after the cut-off time so it won't be shipped untill tomorrow. The cronjob inventory_cleanup_reservations will then run and sees the sum of these reservations is 0 and thus will delete the reservations, resulting in a wrong salable quantity for that product.

### Partially fixed Issues
1. magento/inventory#2311: Improve reservations clean up cron job

### Manual testing scenarios (*)
See description.

### Questions or comments
This is the implementation we are currently using. It doesn't fix magento/inventory#2311 completely but might help. It will clean reservations based on the `object_id` saved in the `metadata` column.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ishakhsuvarov** [comment]
@magento run Integration Tests build

File: app/code/Magento/InventoryReservations/Model/ResourceModel/CleanupReservations.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ public function execute(): void
                 $reservationTable,
                 ['GROUP_CONCAT(' . ReservationInterface::RESERVATION_ID . ')']
             )
-            ->group([ReservationInterface::STOCK_ID, ReservationInterface::SKU])
+            ->group("JSON_EXTRACT(metadata, '$.object_id')")
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryReservations/Model/ResourceModel/CleanupReservations.php]: It's really unfortunate, but Magento still [supports](https://devdocs.magento.com/guides/v2.3/install-gde/system-requirements-tech.html#database) MySQL versions as old as 5.6, which do not have any of the JSON functions. 

We would need to come up with a more versatile solution here to get it merged in the core

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
Hi @JeroenVanLeusden , due to changes in the repository structure (Magento core code is extracted from the MSI code base) we have changed the base branch to `1.1-develop-legacy`. We will keep this branch for a while to make a smooth transition, in the meanwhile, please, apply changes from the current PR to the new code base structure in `1.1-develop` branch. 

[Installation Guide](https://github.com/magento/inventory/wiki/Metapackage-Installation-Guide) may be helpful for local project development. 

Let me know if you need help or have questions!
Thank you for your collaboration!

**Comment by sidolov** [comment]
Hi @JeroenVanLeusden , I have good news! We are going to eliminate MySQL 5.6 support in the Magento 2.4.0, so, you may create PR with changes to the 1.2-develop branch and we will be able to deliver it. Thank you!

**Pull request closed without merging**